File: frontend/src/api/coverage.ts
/**
 * frontend/src/api/coverage.ts
 * API calls for Data Coverage, Pipeline Health, and Database Verification.
 */

import { apiClient } from './client';

export interface FieldFillRate {
  field_name: string;
  populated_count: number;
  total_count: number;
  fill_percentage: number;
}

export interface IngestionRunSummary {
  id: number;
  started_at: string | null;
  completed_at: string | null;
  status: string;
  apps_requested: number;
  apps_processed: number;
  apps_succeeded: number;
  apps_failed: number;
  apps_skipped: number;
}

export interface DataCoverageResponse {
  master_frontier_total: number;
  canonical_apps_in_db: number;
  rejected_inactive_apps: number;
  unaccounted_apps: number;
  reconciliation_status: string;
  total_categories: number;
  total_app_category_links: number;
  total_pricing_plans: number;
  apps_with_pricing: number;
  pricing_coverage_percentage: number;
  total_merchant_reviews: number;
  priority_apps_with_reviews: number;
  field_fill_rates: FieldFillRate[];
  recent_ingestion_runs: IngestionRunSummary[];
  integrity_status: Record<string, boolean>;
}

export interface HealthResponse {
  status: string;
  database_connected: boolean;
  postgres_version: string | null;
  timestamp: string;
}

export async function fetchDataCoverage(): Promise<DataCoverageResponse> {
  return apiClient<DataCoverageResponse>('/coverage');
}

export async function fetchHealth(): Promise<HealthResponse> {
  return apiClient<HealthResponse>('/health');
}

File: backend/schemas/coverage.py
"""
backend/schemas/coverage.py
---------------------------
Pydantic models for Data Coverage and System Health metrics.
"""

from __future__ import annotations

from datetime import datetime
from pydantic import BaseModel, Field


class FieldFillRate(BaseModel):
    field_name: str
    populated_count: int
    total_count: int
    fill_percentage: float


class IngestionRunSummary(BaseModel):
    id: int
    started_at: datetime | None = None
    completed_at: datetime | None = None
    status: str
    apps_requested: int
    apps_processed: int
    apps_succeeded: int
    apps_failed: int
    apps_skipped: int


class DataCoverageResponse(BaseModel):
    master_frontier_total: int
    canonical_apps_in_db: int
    rejected_inactive_apps: int
    unaccounted_apps: int
    reconciliation_status: str

    total_categories: int
    total_app_category_links: int
    total_pricing_plans: int
    apps_with_pricing: int
    pricing_coverage_percentage: float

    total_merchant_reviews: int
    priority_apps_with_reviews: int

    field_fill_rates: list[FieldFillRate]
    recent_ingestion_runs: list[IngestionRunSummary]
    integrity_status: dict[str, bool]


class HealthResponse(BaseModel):
    status: str = "healthy"
    database_connected: bool
    postgres_version: str | None = None
    timestamp: datetime
